File: codes/lua/008_operator_ternary.lua
print('\n-- Ternary Operator in Lua')

function sprint(...)
    local rest_arguments = {...}
    local new_array = {}
    for _, argument in ipairs(rest_arguments) do
        table.insert(new_array, tostring(argument))
    end
    print(table.concat(new_array, ""))
end

function ternary(true_condition, callback_function_if_condition_true, callback_function_if_condition_false)
    if (true_condition == true) then return callback_function_if_condition_true() end
    return callback_function_if_condition_false()
end

CORRECT_ANSWER = "foo"
sprint("correct answer: ", CORRECT_ANSWER)

print("-- using Ternary Operator function \"ternary\"")

my_answer = "bar"
sprint("my answer: ", my_answer)
sprint("is my answer correct: ", ternary((my_answer == CORRECT_ANSWER), (function() return ("correct!") end), (function() return ("wrong!") end)))
-- is my answer correct: wrong!

my_answer = "foo"
sprint("my answer: ", my_answer)
sprint("is my answer correct: ", ternary((my_answer == CORRECT_ANSWER), (function() return ("correct!") end), (function() return ("wrong!") end)))
-- is my answer correct: correct!

print("-- using \"and\" \"or\" operator, the Lua way")

my_answer = "bar"
sprint("my answer: ", my_answer)
sprint("is my answer correct: ", ((my_answer == CORRECT_ANSWER) and "correct!") or "wrong!")
-- is my answer correct: wrong!

my_answer = "foo"
sprint("my answer: ", my_answer)
sprint("is my answer correct: ", ((my_answer == CORRECT_ANSWER) and "correct!") or "wrong!")
-- is my answer correct: correct!

File: codes/r/008_operator_ternary.r
cat("\n# Ternary Operator in R\n")

cat("# using R built-in operator \"if else\"\n")

CORRECT_ANSWER <- "foo"
cat(paste(sep = "", "correct answer: ", CORRECT_ANSWER, "\n"))

my_answer <- "bar"
cat(paste(sep = "", "my answer: ", my_answer, "\n"))
cat(paste(sep = "", "is my answer correct: ", (if (my_answer == CORRECT_ANSWER) "correct!" else "wrong!"), "\n"))
# is my answer correct: wrong!

my_answer <- "foo"
cat(paste(sep = "", "my answer: ", my_answer, "\n"))
cat(paste(sep = "", "is my answer correct: ", (if (my_answer == CORRECT_ANSWER) "correct!" else "wrong!"), "\n"))
# is my answer correct: correct!

cat("# using R built-in function \"ifelse()\"\n")

my_answer <- "bar"
cat(paste(sep = "", "my answer: ", my_answer, "\n"))
cat(paste(sep = "", "is my answer correct: ", (ifelse((my_answer == CORRECT_ANSWER), "correct!", "wrong!")), "\n"))
# is my answer correct: wrong!

my_answer <- "foo"
cat(paste(sep = "", "my answer: ", my_answer, "\n"))
cat(paste(sep = "", "is my answer correct: ", (ifelse((my_answer == CORRECT_ANSWER), "correct!", "wrong!")), "\n"))
# is my answer correct: correct!
